File: apps/api/pi_dash/ee/authentication/desktop.py
# Copyright (c) 2023-present Pi Dash Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

"""CE seam for "did this request come from the Pi Dash desktop app?".

Three endpoints are desktop-only — machine enrollment, the agent model profile,
and the agent model token. Each hands the caller something a browser tab has no
business holding, so "is authenticated" is not a strong enough gate.

CE marks desktop sessions with a session key, because the open-source build
authenticates with Django sessions. The cloud overlay replaces this module to
read the ``client`` claim its OIDC exchange stamps on the desktop access token,
which is the stronger check (a claim cannot be set by a later request).
"""

from __future__ import annotations

#: Session key CE uses to remember that this session was opened by the desktop
#: app. Set at sign-in by the desktop flow; absent for browser sessions.
DESKTOP_SESSION_KEY = "pidash_client"
DESKTOP_CLIENT = "desktop"


def request_is_desktop(request) -> bool:
    """True when ``request`` belongs to a Pi Dash desktop session."""
    session = getattr(request, "session", None)
    if session is None:
        return False
    try:
        return session.get(DESKTOP_SESSION_KEY) == DESKTOP_CLIENT
    except Exception:  # noqa: BLE001 — a broken session store is not a desktop
        return False

File: rust-api/crates/auth/src/permissions/desktop.rs
#![forbid(unsafe_code)]

//! Desktop-session gate (`managed_runner/permissions.py::IsDesktopSession`
//! over `ee/authentication/desktop.py::request_is_desktop`).
//!
//! Three provisioning endpoints are desktop-only (machine enrollment, the
//! agent model profile, the agent model token): "is authenticated" is not a
//! strong enough gate for handing a browser tab those secrets. CE marks
//! desktop sessions with the session key `pidash_client = "desktop"` at
//! sign-in; the cloud overlay replaces the check with the OIDC `client`
//! claim (an F-10 seam). A missing session, a missing key, or a broken
//! session store is "not desktop" — the broad `except` in Python fails
//! closed here too.

/// Session key CE uses to remember a desktop session.
pub const DESKTOP_SESSION_KEY: &str = "pidash_client";
/// Session value marking the desktop app.
pub const DESKTOP_CLIENT: &str = "desktop";

/// Mirror of `request_is_desktop`: `session_value` is the stored
/// `DESKTOP_SESSION_KEY` value (`None` when the session is missing, the key
/// absent, or the store broken).
pub fn is_desktop_session(authenticated: bool, session_value: Option<&str>) -> bool {
    if !authenticated {
        return false;
    }
    matches!(session_value, Some(value) if value == DESKTOP_CLIENT)
}

/// Overlay seam for `ee/authentication/desktop.py::request_is_desktop`.
///
/// The inputs are data, not a request: the caller extracts whatever its
/// build's check needs (the CE session value, the cloud OIDC `client`
/// claim) and the gate decides. The default implementation is the CE
/// session-key check; the private crate replaces the gate, never the
/// callers.
pub trait DesktopGate {
    fn is_desktop(&self, authenticated: bool, session_value: Option<&str>) -> bool {
        is_desktop_session(authenticated, session_value)
    }
}

/// CE gate: the session-key check, unchanged.
pub struct SessionDesktopGate;

impl DesktopGate for SessionDesktopGate {}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn only_desktop_value_passes() {
        assert!(is_desktop_session(true, Some("desktop")));
        assert!(!is_desktop_session(true, Some("browser")));
        assert!(!is_desktop_session(true, Some("")));
        assert!(!is_desktop_session(true, None));
    }

    #[test]
    fn anonymous_never_desktop() {
        assert!(!is_desktop_session(false, Some("desktop")));
    }

    #[test]
    fn default_gate_is_the_session_check() {
        let gate = SessionDesktopGate;
        assert!(gate.is_desktop(true, Some("desktop")));
        assert!(!gate.is_desktop(true, None));
        assert!(!gate.is_desktop(false, Some("desktop")));
    }

    /// The overlay replaces the gate (e.g. with the OIDC `client` claim);
    /// callers behind the trait see the replacement without changing.
    struct ClaimDesktopGate;

    impl DesktopGate for ClaimDesktopGate {
        fn is_desktop(&self, authenticated: bool, session_value: Option<&str>) -> bool {
            authenticated && matches!(session_value, Some("desktop-oidc"))
        }
    }

    #[test]
    fn replacement_gate_wins_behind_the_trait() {
        let gate = ClaimDesktopGate;
        assert!(gate.is_desktop(true, Some("desktop-oidc")));
        assert!(!gate.is_desktop(true, Some("desktop")));
    }
}
